### ISSUE TITLE
Miss column in query builder result for MongoDB

### ISSUE THREAD

**Issue by Densol92** [metabase/metabase#4946]
Metabase version 0.23.1, Running in Docker
Reproduced in Chrome, Firefox, Windows 8, Fedora 23
Metabase internal database - default

I have mongo for store some reports. It looks like:
 {'report_name': 'SOME_REPORT_NAME', 'date': REPORT_DATE, 'some_field': FIELD_VALUE} etc.
I add some fields and everything looks great until today.
I add new report `{ "_id" : ObjectId(SOME_ID), "date" : ISODate("2017-05-01T21:27:47.255Z"), "report_name" : "new_os_ext_count", "new_os_ext_count" : 451 }` and want to make a question in Metabase, and i get this
![selection_063](https://cloud.githubusercontent.com/assets/16528043/25648753/e0db38d2-3001-11e7-9c41-de94da6bf6f5.png)
 as u see "new_os_ext_count"  is missing. If i search directly by id its also missing.
Default query builder is
`[{"$project":{"total":"$total","date~~~default":{"$let":{"vars":{"field":"$date"},"in":"$$field"}},"ui_type":"$ui_type","country":"$country","ppa_servers":"$ppa_servers","servers":"$servers","count":"$count","report_name":"$report_name","extension":"$extension","script":"$script","domains":"$domains","installed_component":"$installed_component","ppa_domains":"$ppa_domains","operation_system":"$operation_system","version":"$version","_id":"$_id","report":"$report","instances":"$instances","language":"$language","client":"$client","php_handler":"$php_handler","installation_time_sec":"$installation_time_sec"}},{"$match":{"report_name":{"$eq":"new_os_ext_count"}}},{"$project":{"total":"$total","date~~~default":"$date~~~default","ui_type":"$ui_type","country":"$country","ppa_servers":"$ppa_servers","servers":"$servers","count":"$count","report_name":"$report_name","extension":"$extension","script":"$script","domains":"$domains","installed_component":"$installed_component","ppa_domains":"$ppa_domains","operation_system":"$operation_system","version":"$version","_id":"$_id","report":"$report","instances":"$instances","language":"$language","client":"$client","php_handler":"$php_handler","installation_time_sec":"$installation_time_sec"}},{"$limit":2000}]`

If i use custom query it works fine `[{"$match":{"report_name":{"$eq":"new_os_ext_count"}}},{"$limit":2000}]`
![selection_064](https://cloud.githubusercontent.com/assets/16528043/25649843/590ffbe0-300b-11e7-8046-27cea2edd090.png)

 How can it be fixed? Not all people know NoSQL

**Comment by salsakran** [comment]
We don't show all columns for every row. We expect the documents in a Mongo collection to all effectively have the same schema.

If you have a per-collection specific key, we won't find and display most of those. In general, I'd recommend naming all the columns the same. Is there a reason you don't use "report" instead of the value of "report" as the key?

**Comment by Densol92** [comment]
My first structure was 
- REPORT_NAME_1
- - DATE_1
- - - SOME_DATA_1
- - - SOME_DATA_2
- - DATE_2
- - - SOME_DATA_1
- - - SOME_DATA_2
- REPORT_NAME_2 ...

But query builder not recognize nested fields - #1328 , #4098 so i still need to do custom query on my own.

**Comment by theage** [comment]
This behavior becomes an issue if you e.g. aggregate into documents where not all documents have all the same properties. If the first document is missing them, there is no way to make them visible other than adding those properties to all results (or at least the first returned one).

It would be nice if we could add custom properties / columns in the column selection dialog, e.g. via text box.

**Comment by priyankinfinnov** [comment]
Cant, we simply get a 'add field' option in 'Data Model' section for any collection.
So, that if we have a key 'aaa' (present in only a few docs, not all) and metabase did not find it while the scanning the schema, we can manually add that and then on metabase can use/show that field.

**Comment by jtlz2** [comment]
@priyankinfinnov Couple of comments:

(i) Isn't one point of mongodb that documents don't have to follow the same schema? So having metabase cope in some way with different schema in the same table would be extremely helpful. At the moment when I follow the workaround I still get just IDs in the table and no other projected columns.

(ii) If you are saying one needs custom functions to mitigate (i), where can one define these so metabase can access them?

**Comment by Nepoxx** [comment]
This could be resolved with a solution like this:
https://github.com/metabase/metabase/issues/10484

@jtlz2 Point (i) is absolutely right;

**Comment by Nidroide** [comment]
The problem appears to me when converting a simple question in a native one.
The simple:

![image](https://user-images.githubusercontent.com/6850004/79109957-b6d2bb00-7d79-11ea-9ed0-30ceb814d35d.png)

All documents share the same schema but not all fields are always present. Converting the question in its native format (no further edit of pipeline) some columns disappear:

![image](https://user-images.githubusercontent.com/6850004/79110451-b850b300-7d7a-11ea-99f5-7c3d49132ff8.png)

As already pointed when values are missing from the first document columns are not generated even if they are all present in the metadata model.

As a partial workaround I can add an $addFields pipeline like that:
```
$addFields: {
	        isolamento_domiciliare: { $ifNull: ['$isolamento_domiciliare', undefined]},
	        ricoverati_altre_province: { $ifNull: ['$ricoverati_altre_province', undefined]},
	        ricoverati_altre_strutture: { $ifNull: ['$ricoverati_altre_strutture', undefined]},
	        ricoverati_sanpio: { $ifNull: ['$ricoverati_sanpio', undefined]},
	        totale: { $ifNull: ['$totale', undefined]},
	        
	    }
```

and all columns appears again, anyway I cannot save the question:

```
metabase           | 04-13 09:57:19 DEBUG middleware.log :: POST /api/card 400 4,0 ms (0 DB calls) 
metabase           | {:errors {:result_metadata "il valore potrebbe essere 0, o se non 0, value must be an array of valid results column metadata maps."}}
``` 

Version: v0.35.2

**Comment by camsaul** [comment]
Since this is a product of how we sync MongoDB columns, I'm going to close it in favor of the other open issues that would address the core problem.

- #10484
- #15576
- #27045

We are currently working on #27045, which will mitigate this problem somewhat. But a complete fix would likely require either #10484 or #15576

Marking as duplicate of #10484
